dvc/repo/experiments/queue/base.py

```diff
@@ -508,28 +519,74 @@ def _format_new_params_msg(new_params, config_path):
             f"from '{config_path}': {param_list}"
         )
 
-    def _update_params(self, params: dict):
-        """Update experiment params files with the specified values."""
-        from dvc.utils.collections import NewParamsFound, merge_params
+    def _update_params(self, path_params: Iterable[str]):
+        """Update param files with the provided `Hydra Override`_ patterns.
+
+        Args:
+            path_params: List of `Hydra Override`_ patterns, provided via
+                `exp run --set-param`.
+
+        .. _Hydra Override:
+            https://hydra.cc/docs/next/advanced/override_grammar/basic/
+        """
+        from hydra._internal.config_loader_impl import ConfigLoaderImpl
+        from hydra.core.override_parser.overrides_parser import OverridesParser
+        from hydra.errors import (
+            ConfigCompositionException,
+            OverrideParseException,
+        )
+        from hydra.types import RunMode
+        from omegaconf import OmegaConf
+
+        from dvc.utils.collections import (
+            merge_dicts,
+            remove_missing_keys,
+            to_omegaconf,
+        )
         from dvc.utils.serialize import MODIFIERS
 
-        logger.debug("Using experiment params '%s'", params)
+        hydra_errors = (ConfigCompositionException, OverrideParseException)
+
+        path_overrides = defaultdict(list)
+        for path_param in path_params:
+            path, _, override = path_param.partition(":")
+            if not override:
+                override = path
+                path = ParamsDependency.DEFAULT_PARAMS_FILE
+
+            path_overrides[path].append(override)
 
-        for path in params:
+        for path, overrides in path_overrides.items():
             suffix = self.repo.fs.path.suffix(path).lower()
+
             modify_data = MODIFIERS[suffix]
-            with modify_data(path, fs=self.repo.fs) as data:
+            with modify_data(path, fs=self.repo.fs) as original_data:
                 try:
-                    merge_params(data, params[path], allow_new=False)
-                except NewParamsFound as e:
-                    msg = self._format_new_params_msg(e.new_params, path)
-                    raise MissingParamsError(msg)
+                    parser = OverridesParser.create()
+                    parsed = parser.parse_overrides(overrides=overrides)
+                    ConfigLoaderImpl.validate_sweep_overrides_legal(
+                        parsed, run_mode=RunMode.RUN, from_shell=True
+                    )
+
+                    new_data = OmegaConf.create(to_omegaconf(original_data))
```

Comment:
Hitting an error right away with floats because OmegaConf can't handle `ruamel.yaml.scalarfloat.ScalarFloat`. Why is `original_data` being passed here?

```
$ dvc exp run -v --set-param lr=0.1
2022-07-29 10:03:20,650 ERROR: unexpected error - Value 'ScalarFloat' is not a supported primitive type
    full_key: lr
    object_type=dict
------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dave/Code/dvc/dvc/cli/__init__.py", line 185, in main
    ret = cmd.do_run()
  File "/Users/dave/Code/dvc/dvc/cli/command.py", line 22, in do_run
    return self.run()
  File "/Users/dave/Code/dvc/dvc/commands/experiments/run.py", line 32, in run
    results = self.repo.experiments.run(
  File "/Users/dave/Code/dvc/dvc/repo/experiments/__init__.py", line 519, in run
    return run(self.repo, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/__init__.py", line 48, in wrapper
    return f(repo, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/run.py", line 45, in run
    return repo.experiments.reproduce_one(
  File "/Users/dave/Code/dvc/dvc/repo/experiments/__init__.py", line 119, in reproduce_one
    self.queue_one(exp_queue, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/__init__.py", line 156, in queue_one
    return self.new(
  File "/Users/dave/Code/dvc/dvc/repo/experiments/utils.py", line 40, in wrapper
    return f(exp, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/__init__.py", line 269, in new
    return queue.put(*args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/queue/workspace.py", line 27, in put
    return self._stash_exp(*args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/queue/base.py", line 343, in _stash_exp
    self._update_params(params)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/queue/base.py", line 571, in _update_params
    new_data = OmegaConf.create(to_omegaconf(original_data))
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/omegaconf.py", line 176, in create
    return OmegaConf._create_impl(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/omegaconf.py", line 882, in _create_impl
    format_and_raise(node=None, key=None, value=None, msg=str(e), cause=e)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 793, in format_and_raise
    _raise(ex, cause)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 771, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set env var OC_CAUSE=1 for full trace
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/omegaconf.py", line 843, in _create_impl
    return DictConfig(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 111, in __init__
    format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 793, in format_and_raise
    _raise(ex, cause)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 771, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set env var OC_CAUSE=1 for full trace
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 109, in __init__
    self._set_value(content, flags=flags)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 649, in _set_value
    raise e
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 646, in _set_value
    self._set_value_impl(value, flags)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 692, in _set_value_impl
    self.__setitem__(k, v)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 316, in __setitem__
    self._format_and_raise(key=key, value=value, cause=e)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/base.py", line 231, in _format_and_raise
    format_and_raise(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 793, in format_and_raise
    _raise(ex, cause)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 771, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set env var OC_CAUSE=1 for full trace
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 310, in __setitem__
    self.__set_impl(key=key, value=value)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/dictconfig.py", line 320, in __set_impl
    self._set_item_impl(key, value)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/basecontainer.py", line 618, in _set_item_impl
    self._wrap_value_and_set(key, value, target_type_hint)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/basecontainer.py", line 634, in _wrap_value_and_set
    self._format_and_raise(key=key, value=val, cause=e)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/base.py", line 231, in _format_and_raise
    format_and_raise(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 873, in format_and_raise
    _raise(ex, cause)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/_utils.py", line 771, in _raise
    raise ex.with_traceback(sys.exc_info()[2])  # set env var OC_CAUSE=1 for full trace
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/basecontainer.py", line 626, in _wrap_value_and_set
    wrapped = _maybe_wrap(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/omegaconf.py", line 1070, in _maybe_wrap
    return _node_wrap(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/omegaconf.py", line 1027, in _node_wrap
    node = AnyNode(value=value, key=key, parent=parent)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/nodes.py", line 135, in __init__
    super().__init__(
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/nodes.py", line 29, in __init__
    self._set_value(value)  # lgtm [py/init-calls-subclass]
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/nodes.py", line 46, in _set_value
    self._val = self.validate_and_convert(value)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/nodes.py", line 76, in validate_and_convert
    return self._validate_and_convert_impl(value)
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/omegaconf/nodes.py", line 154, in _validate_and_convert_impl
    raise UnsupportedValueType(
omegaconf.errors.UnsupportedValueType: Value 'ScalarFloat' is not a supported primitive type
    full_key: lr
    object_type=dict
```

Reply:
> Hitting an error right away with floats because OmegaConf can't handle ruamel.yaml.scalarfloat.ScalarFloat. Why is original_data being passed here?

In order to parse and apply the Hydra override syntax, an OmegaConf instance is required as input.
 
When we try to preserve comments the parsers return custom classes (comment here https://github.com/iterative/dvc/pull/8067#discussion_r933007353) 
I added `to_omegaconf` util to convert custom containers (`CommentedMap`, `CommentedSeq`) to primitives (`dict`, `list`). 
Need to also handle this ScalarFloat class .

Reply:
Should be fixed now. Passed `allow_objects=True` flag to OmegaConf create. This didn't work for containers so the `to_omegaconf` is still needed but it seems to work for non-container custom classes like ScalarFloat